gnovm/pkg/gnolang/alloc.go

```diff
@@ -104,10 +114,90 @@ func (alloc *Allocator) Allocate(size int64) {
 
 	alloc.bytes += size
 	if alloc.bytes > alloc.maxBytes {
-		panic("allocation limit exceeded")
+		if alloc.heap != nil {
+			deleted := alloc.heap.MarkAndSweep()
+			alloc.DeallocDeleted(deleted)
+
+			v, _ := mem.VirtualMemory()
+			ca := uint64(size) > v.Available
+
+			if (alloc.detonate || ca) && alloc.bytes > alloc.maxBytes {
+				panic("allocation limit exceeded")
+			}
+			alloc.detonate = alloc.bytes > alloc.maxBytes
+		}
+	}
+}
+
+func (alloc *Allocator) DeallocDeleted(objs []*GcObj) {
+	for _, obj := range objs {
+		alloc.DeallocObj(obj.tv)
+	}
+}
+
+func (alloc *Allocator) AllocateObj(tv TypedValue) {
+	switch v := tv.V.(type) {
+	case PointerValue:
+		alloc.AllocateType()
+
+		if v.TV != nil {
+			alloc.AllocateObj(*v.TV)
+		}
+	case *StructValue:
+		alloc.AllocateStruct()
+		alloc.AllocateStructFields(int64(len(v.Fields)))
+		alloc.AllocateType()
+		alloc.AllocateHeapItem()
```

Comment:
StructValue does not imply `heapItemValue` allocation, IMO.

Reply:
This code refers to StructValue as a result of a RefExpr.
The VM currently uses `NewHeapItem` for it.

```go
	case *CompositeLitExpr: // for *RefExpr
		tv := *m.PopValue()
		hv := m.Alloc.NewHeapItem(tv)
		return PointerValue{
			TV:    &hv.Value,
			Base:  hv,
			Index: 0,
		}
```